### PR TITLE
fix: support idempotency for framework and spark processors

### PR DESCRIPTION
**Pull request by brockwade633**
*Issue #, if available:*
https://github.com/aws/sagemaker-python-sdk/issues/3451

*Description of changes:*
These changes address a bug described in more detail in the above GH issue. Provides idempotency support for FrameworkProcessors, PySparkProcessors, and SparkJarProcessors.

*Testing done:*
- Extended processing, training, transform, and tuning step unit test suites to include assertions for idempotency
- Verified no duplicate inputs in the pipeline definition from Jupyter notebook that exposed this issue

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by nmadan — APPROVED** [review]

File: src/sagemaker/spark/processing.py [hunk 1]
```diff
@@ -940,9 +940,18 @@ def _extend_processing_args(self, inputs, outputs, **kwargs):
             outputs: Processing outputs.
             kwargs: Additional keyword arguments passed to `super()`.
         """
+
+        if inputs is None:
+            inputs = []
+
+        # make a shallow copy of user inputs
```
Comment (mufaddal-rohawala) [inline src/sagemaker/spark/processing.py:947]: nit: can use the library - https://docs.python.org/3/library/copy.html in future.

**Review by mufaddal-rohawala — CHANGES_REQUESTED** [review]
Units are failing for the changes in the PR @brockwade633 can you fix?

**Comment by brockwade633** [comment]
> Units are failing for the changes in the PR @brockwade633 can you fix?

@mufaddal-rohawala unit tests are fixed, can you please review again?

**Review by mufaddal-rohawala — APPROVED** [review]

**Comment by mufaddal-rohawala** [comment]
Awating reinvent release to fix the failing integ tests.

**Review by mufaddal-rohawala — COMMENTED** [review]
/bot run unit-tests, slow-tests

**Comment by mufaddal-rohawala** [comment]
@brockwade633 can you check the failing units.

File: src/sagemaker/spark/processing.py (hunk 1 above)
Comment (qidewenwhen) [reply src/sagemaker/spark/processing.py:960]: +1 on using this library as the shallow copy code block repeats in multiple places

**Review by qidewenwhen — APPROVED** [review]

**Review by knikure — COMMENTED** [review]
/bot run notebook-tests

**Pull request merged**